### ISSUE TITLE
Can't use callable as lax.scan carry

### ISSUE THREAD

**Issue by GJBoth** [google/jax#6539]
Hi,

I'm working on some code where I need the (vector-valued) gradients of a function up to a certain order. I figured I could do this nicely with `jax.lax.scan`, where the carry is the gradient function and the save that function evaluated at a certain point, i.e.

```
def deriv(f, x):
    """ Returns grad function and grad evaluated at x."""
    y, vjp_fn = jax.vjp(f, x)
    return lambda x: vjp_fn(x)[0], vjp_fn(jnp.ones(y.shape))[0]
```

This function works when I call it like this,
```
# This works
df_fn, df = deriv(f, x)
d2f_fn, d2f = deriv(df_fn, x)
d3f_fn, d3f = deriv(d2f_fn, x)
```

But I can't use it in a scan:
```
# This fails
du = scan(lambda carry, _: deriv(carry, x), f, None, length=3)
```

I have a minimal example in a notebook [here](https://colab.research.google.com/drive/1860bkMXo0x-7Cqu37Q8WNGi-XYreA65v#scrollTo=FY-kTy2hx2XR). 

Is this intended behaviour? Or a limitation on XLA's part? Or should I use a different tactic?

**Comment by NeilGirdhar** [comment]
I'm not on the JAX team, but I'm pretty sure the problem is that the carry has to be a PyTree, and a function is not (normally) a PyTree.  (Your code is a neat idea though.)

**Comment by shoyer** [comment]
Next time please include the full error message in your bug report :)

Neil is right in identifying that you can't pass callables into `carry` unless you make them pytrees. But there's an even bigger issue here. By design `scan` needs to run the same compiled code in each iteration. JAX's autodiff happens before compiling, so even if you turned the call to `deriv` into a PyTree, I'm pretty sure you you would end up with an error from inside `jax.vjp`.

I agree with Neil's suggest to use a regular Python for-loop. That would result in maximally efficient code in this case.

By the way, if you need very high order derivatives, you might find `jax.experimental.jet` to be useful.

**Comment by GJBoth** [comment]
Yeah, I should've - will do next time. I tried to see how far I could take my code, but in the end I indeed settled on a for loop. Thanks Neil and Stephan.

@shoyer Jet is not very well documented - is the API going to change, is it going to be removed or did you just not get around to it yet? I need 5th order derivatives, which really slows things down, so jet seems like a nice solution.

**Comment by mattjj** [comment]
> Jet is not very well documented - is the API going to change, is it going to be removed or did you just not get around to it yet? I need 5th order derivatives, which really slows things down, so jet seems like a nice solution.

The API likely won't change, and it won't be removed. When we put it into `jax` rather than `jax.experimental` we'll make sure it's documented (that's the main blocker actually), but as you say we haven't gotten around to it yet.

**Comment by froystig** [comment]
We have an issue open for that: #3730

**Comment by GJBoth** [comment]
Thanks, looks interesting, I'll have a look!
